tests/integration/test_charm.py

```diff
@@ -1,49 +1,137 @@
 # Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-import pytest
-from ops.model import ActiveStatus
+import json
+import logging
+
+import ops
+import requests
+from boto3 import client
+from botocore.config import Config
+from ops.model import Application
 from pytest_operator.plugin import OpsTest
 
+logger = logging.getLogger(__name__)
+
+
+async def test_workload_online_default(
+    ops_test: OpsTest,
+    app: Application,
+):
+    unit_informations = json.loads(
+        (await ops_test.juju("show-unit", app.units[0].name, "--format", "json"))[1]
+    )
+    mattermost_ip = unit_informations[app.units[0].name]["address"]
+    response = requests.get(f"http://{mattermost_ip}:8065", timeout=5)
+    assert response.status_code == 200
+    assert "Mattermost" in response.text
+
+
+async def test_s3_storage(
+    ops_test: OpsTest,
+    model: ops.model.Model,
+    app: Application,
+    localstack_s3_config: dict,
+    test_user: dict,
+    tmp_path,
+):
+    """
+    arrange: after charm deployed and openstack swift server ready.
+    act: update charm configuration for openstack object storage plugin.
+    assert: a file should be uploaded to the openstack server and be accessible through it.
+    """
+
+    await app.set_config(
+        {
+            "s3_enabled": "true",
+            "s3_endpoint": localstack_s3_config["endpoint_without_scheme"],
+            "s3_bucket": localstack_s3_config["bucket"],
+            "s3_region": localstack_s3_config["region"],
+            "s3_access_key_id": localstack_s3_config["credentials"]["access-key"],
+            "s3_secret_access_key": localstack_s3_config["credentials"]["secret-key"],
+            "s3_server_side_encryption": "false",
+            "extra_env": '{"MM_FILESETTINGS_AMAZONS3SSL": "false","MM_SERVICESETTINGS_ENABLELOCALMODE": "true","MM_SERVICESETTINGS_LOCALMODESOCKETLOCATION": "/tmp/mattermost.socket"}',
+        }
+    )
+    await model.wait_for_idle(status="active", raise_on_error=False)
+
+    unit_informations = json.loads(
+        (await ops_test.juju("show-unit", app.units[0].name, "--format", "json"))[1]
+    )
+    mattermost_ip = unit_informations[app.units[0].name]["address"]
```

Comment:
I think the `mattermost_ip` fixture can be used here as well :)

Reply:
same answer